Source/DafnyCore/Auditor/Assumption.cs

```diff
@@ -0,0 +1,211 @@
+/// </summary>
+public struct AssumptionProperties {
+  internal bool IsGhost;
+  internal bool IsSubsetType;
+  internal bool IsCallable;
+  internal bool IsTrait;
+  internal bool MissingBody;
+  internal bool HasAxiomAttribute;
+  internal bool HasExternAttribute;
+  internal bool HasVerifyFalseAttribute;
+  internal bool HasAssumeInBody;
+  internal bool MayNotTerminate;
+  internal IEnumerable<AttributedExpression> RequiresClauses; // non-null implies one or more elements
+  internal IEnumerable<AttributedExpression> EnsuresClauses; // non-null implies one or more elements
+
+  /// <summary>
+  /// Determines whether the set of properties constitutes an assumption.
+  /// </summary>
+  /// <returns>whether this entity contains assumptions</returns>
+  public bool HasAssumptions() {
+    return HasAxiomAttribute ||
+           (IsCallable &&
+            (((EnsuresClauses is not null || RequiresClauses is not null) &&
+              (MissingBody || HasExternAttribute)) ||
+             HasAssumeInBody)) ||
+           (MayNotTerminate && (IsCallable || IsTrait));
+  }
+
+  /// <summary>
+  /// Describe all of the assumptions this entity makes.
+  /// </summary>
+  /// <returns>a list of (assumption, mitigation) pairs</returns>
+  public IEnumerable<(string, string)> Description() {
+    if (HasAxiomAttribute) {
+      return new List<(string, string)>{
+        ("Declaration has explicit `{:axiom}` attribute.",
+         "Provide a proof, test, or other justification.")
+      };
+    }
+
+    List<(string, string)> descs = new();
+
+    if (IsCallable && MissingBody) {
+      descs.Add(
+        ((IsGhost ? "Ghost" : "Compiled") +
+          " declaration has no body" +
+          (EnsuresClauses is null ? "." : " and has an ensures clause."),
+        "Provide a body or add `{:axiom}`."));
+    }
+    if (IsCallable && HasExternAttribute && RequiresClauses is not null) {
+      descs.Add(
+        ("Declaration with `{:extern}` has a requires clause.",
+         "Test external caller (maybe with `/testContracts`) or justify."));
+    }
+    if (IsCallable && HasExternAttribute && EnsuresClauses is not null) {
+      descs.Add(
+        ("Declaration with `{:extern}` has a ensures clause.",
+         "Test external callee (maybe with `/testContracts`) or justify."));
+    }
+    if (IsCallable && HasAssumeInBody) {
+      descs.Add(
+        ("Definition has `assume` statement in body.",
+         "Replace with `assert` and prove or add `{:axiom}`."));
+    }
+    if (IsCallable && MayNotTerminate) {
+      descs.Add(
+        ("Method may not terminate (uses `decreases *`).",
+         "Provide a valid `decreases` clause or justify the absence."));
+    }
+    if (IsTrait && MayNotTerminate) {
+      descs.Add(
+        ("Trait method calls may not terminate (uses `{:termination false}`).",
+         "Remove or justify."));
+    }
+    if (HasVerifyFalseAttribute) {
+      descs.Add(
+        ("Definition has `{:verify false}` attribute.",
```

Comment:
Every assumption detected here is emitted as a warning by default (but can be written to a file in one of several formats, instead, as an option). So it seems to me the question is more one of which of the assumption types identified here should be warned about when not using the auditor. I could imagine that some of them might always be on, and some might be turned on by something like a `--warn-assumptions` flag. Do either of you have thoughts on that?

The fact that it may make sense to always emit some of these warnings argues even more strongly for including this functionality in the core implementation of Dafny rather than as an optional plugin.

Reply:
> To warn or not to warn, the question is more about when it is appropriate to warn users and how much. It's not debatable to warn the user during an audit. However, I'm not sure I want to clutter my diagnostics with warnings when all I want is the result of verifying a portion of code. So I wouldn't mind these warnings if they were the only thing left (meaning I'm ready to move forward, I fixed all my errors on unrelated code), but while I'm having errors in my IDE, having such warnings for annotations I added manually myself to bypass the verifier would be extremely annoying.

Could you explain what a warning means to you? 

To me a warning means I need to resolve this before I create a commit, so I feel free to ignore them or hide them until I'm happy with my code and I'm ready to polish it for a commit.

It sounds like for you a warning is something that you already want to resolve while you're still getting verification to pass. Why is that?

> Every assumption detected here is emitted as a warning by default

What do you mean? With a warning I mean that if you run `dafny verify` on a source file that includes `{:verify false}`, it returns a warning for that. Right now, that's not the case.

> I could imagine that some of them might always be on

I think I said this before, but any warning that is _not_ on by default is a UX flaw to me. We're telling users that probably they don't want to use this (because the default is OFF), but they should consider using it (that's why we added it). It's better not to overload the user with these small choices and instead just give a recommendation.

`--warn-assumptions` should be on by default.

Reply:
> Could you explain what a warning means to you?
To me a warning means I need to resolve this before I create a commit, so I feel free to ignore them or hide them until I'm happy with my code and I'm ready to polish it for a commit.
It sounds like for you a warning is something that you already want to resolve while you're still getting verification to pass. Why is that?

I don't place the bar of resolving warnings before creating a commit. 
Moving forward one sub-proof is enough for me when writing Dafny to realize I should commit my result, even if there are other assume false. These kinds of warnings should be resolved when creating a PR to the main branch however, as they should be detected by the CI.

I'm not sure of your description of a warning, but it supposes that there is a single notion of warning. But in a verification-aware language like Dafny, I started recently to strongly believe there should be two kinds of warnings: verifier warnings and compiler/resolver warnings. To me, the only reason to trigger a verifier warning is when Dafny was able to prove that the code is correct, but it found unproven assumptions, or obviously false requirements, that the user should be aware of - so that they don't commit their code knowingly.
Users writing Dafny regularly understand why, as they are using proof scaffolding techniques where warnings would not be relevant until they actually prove something. Adding warnings on top of their regular proof authoring experience where they already struggle to prove a goal would be suboptimal.
Of course, these warnings should be reported by the auditor as well.

The second notion of warnings are regular warnings triggered by any mainstream language, such as a redundant case, or a variable that looks like a existing constructor in a pattern matching. These should be reported in any case.

We don't have mechanisms to hide warnings as of today, so I'm not counting on hiding warnings as part of a workflow - and I'm not sure it would be even desirable to force people hiding verification warnings. This looks to me like a UX flaw.

On the side, I do agree that every warning should be turned on by default. But only shown when some conditions are met, as explained above.

Reply:
> I don't place the bar of resolving warnings before creating a commit.

Sorry, I should have said before pushing to CI.